rfd/0080-piv-private-key.md

```diff
@@ -0,0 +1,392 @@
+
+```yaml
+auth_service:
+  ...
+  authentication:
+    ...
+    require_session_mfa: off | on | hardware_key | hardware_key_touch
+```
+```yaml
+kind: cluster_auth_preference
+version: v2
+metadata:
+  name: cluster-auth-preference
+spec:
+  require_session_mfa: off | on | hardware_key | hardware_key_touch
+```
+```yaml
+kind: role
+version: v5
+metadata:
+  name: role-name
+spec:
+  role_options:
+    require_session_mfa: off | on | hardware_key | hardware_key_touch
+```
+
+- `on`: Current per-session MFA functionality - users are required to pass an MFA challenge with a registered MFA device in order to start new SSH|Kubernetes|DB|Desktop sessions. Non-session requests, and app-session requests are not impacted.
+- `hardware_key`: Essentially combines `require_session_mfa: on` with `private_key_policy: hardware_key`. MFA tap is only required for session requests, but all Teleport API requests require certificates backed by a hardware private key.
+- `hardware_key_touch`: This is the same as `private_key_policy: hardware_key_touch`. Per-session MFA is disabled in favor of touch checks directly on the hardware key the user used to login.
+  - This will require some changes to the per-session MFA checks. If a user has `require_session_mfa: hardware_key_touch`, then this should override any other per-session MFA requirement, including the cluster-wide requirement or node/access specific requirements.
+
+This change will require changing the `require_session_mfa` fields above from a `bool` to a `string`. This will be handled by introducing a new proto field and custom marshalling logic to maintain interoperability between new and old servers and clients. See https://github.com/gravitational/teleport/pull/12054 for an example of this.
+
+#### Enforcement - Attestation
+
+In order to enforce user private key usage, we need to take a certificate's public key and tie it back to a trusted hardware device, which can be done with attestation, as explained [above](#attestation).
+
+Attestation will be handled during the normal login/certificate signing process by adding a new `AttestationRequest` field to login requests. For all login paths, we need to include the `AttestationRequest` field in the http request objects:
+
+```go
+// lib/client/weblogin.go
+type SSOLoginConsoleReq struct {
+  ...
+	AttestationRequest AttestationRequest `json:"attestation_request,omitempty"`
+}
+
+type CreateSSHCertReq struct {
+  ...
+	AttestationRequest AttestationRequest `json:"attestation_request,omitempty"`
+}
+
+type AuthenticateSSHUserRequest struct {
+  ...
+	AttestationRequest AttestationRequest `json:"attestation_request,omitempty"`
+}
+```
+
+For SSO login, the `AttestationRequest` field also needs to be added to each SSO auth request type (`OIDCAuthRequest`, `SAMLAuthRequest`, `GithubAuthRequest`), so we will make `AttestationRequest` a proto type.
+
+```proto
+// AttestationRequest is a request to attest an attestation request.
```

Comment:
That's a tough sentence. :)

Could we rename messages to PIVAttestationStatement / YubiKeyPIVAttestationStatement? I think it makes more sense talking about "attestation statements" than "attestation request/response".

Reply:
Yes agreed, this was brought up in the implementation. I'll update this to `AttestationStatement` and `AttestationData`.